### DISCUSSION TITLE
Can't open rtsp feed from ipcam

### DISCUSSION THREAD

**Discussion by acleitao-projects** [ultralytics/ultralytics#8901] [General]
I don't know if this is the right place if it's not please just tell me and I will delete and repost.

A week ago I started to study yolo again. Got an old ipcam and decided to create a surveilance system. After some was able to connect the feed (cgi feed) and detect classes. But that camera is only 640x something resolution and as I like the results I wanted and upgrade. I saw yolov8 can use rtsp so I just bought 4 cameras with that protocol. 

I searched everywhere and couldn't find a way to connect that feed. The cameras are yoosee compatible with rtsp protocol. Mine is on rtsp://admin:admin@192.168.15.6:554/onvif1 it connects on vlc player but I can't make it work with a simple yolo model.predict. This is my code

```
from ultralytics import YOLO
model = YOLO("yolov8n.pt")

results = model.predict(source='rtsp://admin:admin@192.168.15.6:554/onvif1',stream=True, show=True)

print(results)
```

If someone can help will be much appreciated as I'm almost returning the cameras lol

sorry one more edit this is what I get as output
Connected to pydev debugger (build 231.9161.41)
<generator object BasePredictor.stream_inference at 0x0000028D3B08E660>

Process finished with exit code 0

Thx guys

**Comment by glenn-jocher** [comment]
Hello! It looks like you're on the right track with using YOLOv8 for your surveillance system, and using RTSP feeds is definitely a supported feature. Your code snippet seems mostly correct, but let's try a slight modification to properly handle the streaming and results visualization. Since `model.predict()` with `stream=True` returns a generator, you'll need to iterate over it to process and display the frames. Here's how you can do it:

```python
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n.pt")

# Create a generator for the RTSP feed
results = model.predict(source='rtsp://admin:admin @alex-gpt.168.15.6:554/onvif1', stream=True)

# Iterate over the generator to process and display each frame
for result in results:
    result.show()  # Display the frame with detections
```

This modification ensures that each frame fetched from your RTSP feed is processed and displayed. The `result.show()` method within the loop will display the frames with detections. Make sure you have a display environment set up if you're running this on a headless server, or you might not see the output windows.

If you're still encountering issues, it might be worth checking if there are any network restrictions or firewall settings that could be blocking the RTSP stream access for your Python environment, especially since you mentioned it works with VLC player.

Hope this helps, and you don't have to return those cameras! 😄

**Comment by acleitao-projects** [comment]
Hey Glenn thanks for your help. I used your code but still the same. My results object comes back empty and not even get inside the for loop. 
![image](https://github.com/ultralytics/ultralytics/assets/129627122/26dbf273-4381-425a-8954-9f000a9b54bb)

if I print it I get this
<generator object BasePredictor.stream_inference at 0x000001BEDC5D8BA0>

I did a lot of research yesterday and read something(don't really know if it is true) that opencv can't handle h264 compression. I don't really know much about yolo but I guess it uses opencv in the background? I was checking my camera specifications and don't see anything there too (it doesn't say it's h264 but its the most commom compression).

Problem is apart from that doest give me any other error like unable to reach or can't login. And at every other way I try to access the rtsp feed it works. VLC is ok, onvif device manager is ok. I dont think is something related to the cloud service(yoosee) as I'm doing this locally.

I tried to open with opencv using ffmpeg still the same. Do you  have any other ideas?

**Reply by pderrenger** [reply]
Hi there! 😊 It sounds like you're encountering a tricky issue with streaming RTSP feeds into YOLOv8. While YOLOv8 does indeed use OpenCV for video processing, OpenCV should generally handle H264 streams well, especially when compiled with FFmpeg support.

Given that your RTSP feed works in VLC and other applications, the issue might not be directly related to H264 compression but could be related to how the stream is being accessed or processed in your Python environment.

One thing you might try is to explicitly use OpenCV to capture the RTSP stream and then pass the frames to YOLOv8 for prediction. This approach can sometimes offer more control and help isolate issues. Here's a quick example of how you might set this up:

```python
import cv2
from ultralytics import YOLO

# Initialize YOLO model
model = YOLO("yolov8n.pt")

# Open the RTSP stream
cap = cv2.VideoCapture('rtsp://admin:admin @alex-gpt.168.15.6:554/onvif1')

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Predict using YOLOv8
    results = model.predict(frame)
    results.show()  # Display the frame with detections

cap.release()
```

This method ensures that OpenCV handles the stream directly, which might bypass the issue you're encountering. If you're still facing problems, it could be worth checking if your OpenCV build includes FFmpeg support or if there are network-related issues that might be affecting the stream's accessibility to your Python script.

Let's keep troubleshooting, and hopefully, we'll get to the bottom of this! 🚀

**Comment by acleitao-projects** [comment]
Let me just add a bit more context to this. I think is useless info but who knows what can help.

I'm running this in a Windows 11 Acer predator laptop. Still running on CPU as I didn't want to go through the rustle of installing cuda before I can get it running.

I'm using pycharm as my IDE and virtualenv. I have a bunch o packages in this project as I was doing tests trying to figure out why it doesn't work.  I think the most important are
```
Ultralytics 8.1.27
Torch 2.2.1
Python-ffmpeg 2.0.10
Pillow 10.2.0
Python-opencv 4.9.0.80
numpy 1.26.4
```

there are 3 cameras already working two of them are old ipcameras the third is my old phone (hooked up to my 3d print was just a test)
Url for the old cameras look like this
http://192.168.15.20:5000/videostream.cgi?user=admin&password=admin
For the phone looks the same without user and password.

The code you suggested works just fine with any of the three old ones. Here is a frame
![image](https://github.com/ultralytics/ultralytics/assets/129627122/8f0313d1-8ffc-4a2e-bcc4-0b64ea23551a)

but as soon as I try the new camera with the rtsp feed its over

I dont have any links for the camera I'm using in english but with translator I think is possible to understand so. This is my new camera(here is sold under the name orbitronics just a chinese white label camera). 
***Edited:
I think I found a link in english on aliexpress so I think this is the same camera
https://pt.aliexpress.com/item/1005005212861452.html?channel=twinner 

![image](https://github.com/ultralytics/ultralytics/assets/129627122/b5535adc-7019-4d10-9c78-c8171918e117)

https://produto.mercadolivre.com.br/MLB-1574876041-cmera-externa-ip-prova-dagua-infravermelho-externa-wifi-hd-_JM?quantity=3

Its a yoosee camera rtsp for using with nvr is enable.

My last resources will be put a docker up and try to run on a linux server but I dont think that will make any difference.

**Reply by pderrenger** [reply]
Hey there! Thanks for reaching out and providing such detailed context. It really helps to understand your setup better. 😊

From what you've described, it seems like your setup and dependencies are correctly configured, especially since you're able to connect to your older cameras and even a phone camera without issues. The RTSP feed from your new cameras should theoretically work the same way.

One thing to consider is that RTSP streams can sometimes be a bit finicky due to network issues, authentication problems, or codec incompatibilities. Since you mentioned the stream works in VLC, network and authentication are likely fine, but it's worth double-checking the codec compatibility with `python-ffmpeg`.

Here's a small tweak you might want to try, explicitly specifying the `ffmpeg` backend for capturing the stream. This can sometimes help with tricky streams:

```python
from ultralytics import YOLO

# Explicitly specify ffmpeg backend
model = YOLO("yolov8n.pt", backend='ffmpeg')

# Attempt to connect to the RTSP stream
results = model.predict(source='rtsp://admin:admin @alex-gpt.168.15.6:554/onvif1', stream=True, show=True)

# Since results is a generator, you might want to loop through it to process frames
for result in results:
    print(result)
    # Process each frame here
```

If this doesn't resolve the issue, running on a Linux server with Docker could indeed offer a different environment that might handle the stream more gracefully, though it's a bit of a long shot if everything else is configured correctly.

Lastly, codec compatibility is something worth exploring. Ensure that the codecs used by your new cameras are supported by `python-ffmpeg` and that there aren't any known issues with them.

Hope this helps! If you're still stuck, feel free to share any error messages or further details. We're here to help!

**Reply by acleitao-projects** [reply]
Hey there Paula,

I tried this code but looks like Yolo doesn't like the backend argument. Got an error then I checked the implementation and couldn't find anything there too. I will put up a vm and see if it runs on linux... that at least will give me peace about dependencies and configurations. If doesnt work I will return it and try to find another one. I just would like to know what to buy lol trial and error can get expensive lol

**Reply by pderrenger** [reply]
@alex-gpt hey Paula! 😊

Apologies for the confusion earlier. You're right, the `backend` argument isn't available in the current YOLOv8 implementation. My mistake there. Trying out a VM with Linux sounds like a solid plan to isolate the issue further.

Regarding camera selection, it's generally a good idea to look for cameras with well-documented support for RTSP and compatibility with common streaming codecs like H.264. Brands or models with active community forums or tech support can also be a plus, as they might offer helpful insights or solutions.

Best of luck with the VM, and let us know how it goes!

**Comment by acleitao-projects** [comment]
Funny thing.....As I said I started a virtual machine with ubuntu. I ran the quick start guide, try to run my feed and... nothing... BUT... I gave a try to my opencv code using video capture and the image showed up so the camera works just something is wrong. 

I closed the vm cameback to windows and tried to no use venv. Didnt work. I started the vm again... and ... not working again lol. New reboot... working properly. This is the code I'm using for openCV

os.environ["OPENCV_FFMPEG_CAPTURE_OPTIONS"] = "rtsp_transport;0"
video_capture = cv2.VideoCapture("rtsp://admin:nvr76rtsp@192.168.15.6:554/onvif1", cv2.CAP_FFMPEG)
while(1):
    ret, frame = video_capture.read()
    cv2.imshow('Video', frame)
    cv2.waitKey(1)

This line did the trick. 
os.environ["OPENCV_FFMPEG_CAPTURE_OPTIONS"] = "rtsp_transport;0"

As this is a graffiti surveilance system I need to be as most real time as I can so I dont want to use an extra layer adding opencv to capture and then send it to yolo. Is there anyway to add that envirornment flag to yolo.predict? even if I have to fork and make my own branch? If so where can I find the load feed routines?

and try to understand why it doesnt work in my windows computer. I only get [rtsp @ 000002b973152840] Nonmatching transport in server reply from this code

**Reply by pderrenger** [reply]
Hey there! 😊 It's great to hear about your project and your persistence in getting everything to work. It seems like you've pinpointed the issue to the RTSP transport setting when using OpenCV, which is a good find.

For integrating the RTSP transport setting directly into the YOLOv8 `predict` method without the need for OpenCV as an intermediary, you might need to adjust the source code where YOLOv8 handles the video stream input. This would indeed require a bit of customization on your part.

The YOLOv8 model uses PyTorch for its operations, and the streaming functionality is likely handled through a combination of PyTorch utilities and possibly other libraries for video streaming like OpenCV or similar. However, YOLOv8 itself doesn't expose an environment variable similar to `OPENCV_FFMPEG_CAPTURE_OPTIONS` directly.

To implement a similar functionality, you would need to look into the `ultralytics` package, specifically where the video source is initialized and processed. This could be in the file handling the `predict` method or possibly a utility file related to video streaming. Unfortunately, without direct access to the YOLOv8 source code (as it's encapsulated within the `ultralytics` package), I can't provide an exact location.

If you're comfortable with diving into the source code, you could fork the repository, implement a way to specify RTSP transport options (similar to how you did with OpenCV), and then use your forked version for your project.

As for the issue with Windows, it might be related to how FFmpeg is configured or accessed by the underlying libraries YOLOv8 uses for video streaming. Ensuring that FFmpeg is correctly installed and accessible in your Windows environment might help.

Remember, while modifying the source code can give you the flexibility you need, it also means you'll need to maintain your version with any updates from the original repository.

I hope this gives you a starting point for integrating the RTSP transport setting directly into YOLOv8. Good luck with your surveillance system project! 🚀

**Comment by acleitao-projects** [comment]
Ok guys, problem solved!!!!! Thanks for your help.

First RTSP doesn't work with Stream=False as stated in the documents this kind of source should be used with stream=True
Second on linux I only made it work using openCV to capture the frame but I stop testing as soon as I realize was a setup problem and not a code problem.
To fix the problem in windows 11 I just created a envirornment variable called OPENCV_FFMPEG_CAPTURE_OPTIONS and set it to "rtsp_transport;0". I did this in windows control panel not on my code os.environ seems to not work properly

Doing tht and using this code solved the problem:
```
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n.pt")

# Create a generator for the RTSP feed
results = model.predict(source='rtsp://admin:admin @alex-gpt.168.15.6:554/onvif1', stream=True)

# Iterate over the generator to process and display each frame
for result in results:
    result.show()  # Display the frame with detections
```

I hope this post can help people in the near future lol... I waisted too much time on that lol. 

Thank you Glenn and Paula for the guidance!

**Reply by pderrenger** [reply]
Hey there! 🎉 I'm thrilled to hear you've resolved the issue with the RTSP feed and YOLOv8. Your detailed explanation and solution are incredibly valuable, and I'm sure they'll help others facing similar challenges. Setting the `OPENCV_FFMPEG_CAPTURE_OPTIONS` environment variable is a neat trick for Windows users, and using `stream=True` for RTSP sources is indeed crucial. Your perseverance and willingness to share your findings are much appreciated. Thanks for the shoutout to the team, but remember, it's the awesome community, including you, that makes YOLOv8 so great! Keep up the fantastic work, and don't hesitate to reach out if you embark on new adventures with YOLOv8. Happy detecting! 🚀
